### ISSUE TITLE
New line handling and delimiting

### ISSUE THREAD

**Issue by ethan-ou** [fuhrmanator/GIFT-grammar-PEG.js#42]
Continued from ethan-ou/vscode-gift#7

New lines should be handled in the GIFT parser so that \n adds a new line and blank lines delimit questions.

**Comment by didoudiaz** [comment]
IMHO this is more general than \n. I mean, the parser should treat similarly all escaped sequences:  \n \= \{... and so on for all  GIFT meta-character:  ~ = # { }  : 

There are two choices for the parser:

1. recognize them and translated them as 1 character
2. pass them *as is*, ie. as a string of 2 characters. 

Option 1 is more natural for a lexer+parser. However, in this case, it depends on backwards compatibility with existing codes using the parser. How are treated other escape sequences, e.g. how are handled \=, \{, ... ? My advice would be to do the same with \n. If parser handled them as option 1 the parser should (be changed to) do the same for \n. If the parser follows option 2, it should do the same for \n. Depending on this, existing codes using the parser might be updated or not.

**Comment by fuhrmanator** [comment]
@ethan-ou you looked at the PHP code that parses GIFT. Do you remember (know) if it converts `\n` to newline?

**Comment by fuhrmanator** [comment]
OK - I had a look at the GIFT tests for Moodle and couldn't find a test for the processing of `\n` in GIFT, despite the specification.

@didoudiaz the grammar already deals with the escaped characters and we have lots of tests, such as https://github.com/fuhrmanator/GIFT-grammar-PEG.js/blob/master/tests/questions/escapeAll.gift which yields an object specified in https://github.com/fuhrmanator/GIFT-grammar-PEG.js/blob/master/tests/questions/escapeAll.json

I'm looking at https://github.com/moodle/moodle/blob/ea32d651e9f41af8401bbd9fdb701d0554a82971/question/format/gift/format.php to see if I can figure out what's happening. I may just use Moodle to import/export some questions with `\n` and newline to see what's happening.

**Comment by fuhrmanator** [comment]
If I import the following GIFT:
```
::newline in GIFT::blah\nblah
```
and then export it as GIFT, I get back the same format.

However, if I try to edit the question in Moodle, an HTML editor opens and `blah blah` appears (one one line), so it appears the newline is removed. I exported to XML and also there is no newline. I guess the only way to know how Moodle is storing it internally is to look in the database (is it a real newline or a `\n` that gets converted when Moodle displays the question).

I suppose none of this matters for a Javascript version. We only have to decide. That is, previewers or quiz programs using the GIFT parser need to know what to expect (a newline or a `\n`). Ideas?

**Comment by ethan-ou** [comment]
Just wanted to point out some interesting lines in the way GIFT handles line breaks. Hopefully it'll help give some ideas:

At [line 561](https://github.com/moodle/moodle/blob/ea32d651e9f41af8401bbd9fdb701d0554a82971/question/format/gift/format.php#L561) there's a function that converts questions from Moodle to GIFT. The way it handles new lines is particularly interesting:

```php
    protected function repchar($text, $notused = 0) {
        // Escapes 'reserved' characters # = ~ {) :
        // Removes new lines.
        $reserved = array(  '\\',  '#', '=', '~', '{', '}', ':', "\n", "\r");
        $escaped =  array('\\\\', '\#', '\=', '\~', '\{', '\}', '\:', '\n', '');

        $newtext = str_replace($reserved, $escaped, $text);
        return $newtext;
    }
```

Essentially, other characters are escaped but new lines are not. A line feed character `\n` ends up the same when it's processed. Carriage returns however are removed entirely. So if we're working in CRLF, we go from `\r\n` to `\n`.

When writing Moodle questions to GIFT, the [code for question generation](https://github.com/moodle/moodle/blob/ea32d651e9f41af8401bbd9fdb701d0554a82971/question/format/gift/format.php#L690) also uses unescaped line feeds `\n`. And when [processing comments](https://github.com/moodle/moodle/blob/ea32d651e9f41af8401bbd9fdb701d0554a82971/question/format/gift/format.php#L170) into Moodle's internal format, the parser uses raw line feeds:

```php
  foreach ($lines as $key => $line) {
            $line = trim($line);
            if (substr($line, 0, 2) == '//') {
                $comments .= $line . "\n";
                $lines[$key] = ' ';
            }
        }
  $text = trim(implode("\n", $lines));
```

At the same time, escaped line feeds are also supported. GIFT converts them to special characters to avoid parsing errors, before converting back to their unescaped character variants.

```php
protected function escapedchar_pre($string) {
        // Replaces escaped control characters with a placeholder BEFORE processing.

        $escapedcharacters = array("\\:",    "\\#",    "\\=",    "\\{",    "\\}",    "\\~",    "\\n"  );
        $placeholders      = array("&&058;", "&&035;", "&&061;", "&&123;", "&&125;", "&&126;", "&&010");

        $string = str_replace("\\\\", "&&092;", $string);
        $string = str_replace($escapedcharacters, $placeholders, $string);
        $string = str_replace("&&092;", "\\", $string);
        return $string;
    }

    protected function escapedchar_post($string) {
        // Replaces placeholders with corresponding character AFTER processing is done.
        $placeholders = array("&&058;", "&&035;", "&&061;", "&&123;", "&&125;", "&&126;", "&&010");
        $characters   = array(":",     "#",      "=",      "{",      "}",      "~",      "\n"  );
        $string = str_replace($placeholders, $characters, $string);
        return $string;
    }
```

In summary:

* It's likely that Moodle processes line feeds as a single character
* Escaped and unescaped line feeds are both supported in GIFT, but the internal code prefers unescaped line feeds
* Carriage returns are removed entirely from GIFT files

**Comment by ethan-ou** [comment]
@fuhrmanator to answer your point about the HTML editor, I would take a guess and say when opening the HTML editor, Moodle does internal processing to remove new lines before displaying the question to you.

Since the parsing code for GIFT actually injects line feeds `\n` into the final output, (and because there's no code in Moodle's GIFT parser that converts an internal new line representation to line feeds), line feeds probably pass through to Moodle's database.

**Comment by fuhrmanator** [comment]
@ethan-ou I pushed a new branch to [here](fuhrmanator/GIFT-grammar-PEG.js/tree/explicit-parse-newlines). The tests pass in Windows and WSL (Ubuntu). Now, instead of `\n` literal passing into text, there is a newline character. So, previewers will have to change. I have not updated the online previewer yet.

**Comment by ethan-ou** [comment]
@fuhrmanator checked your tests. This means basically we'll still be handling \n in the preview side right? Or is the assumption that new lines characters will automatically break in the preview?

Also I just realised something thanks to your tests. New lines would be handled the same regardless of format (nothing in the moodle code suggests there's any parsing differences), so we should probably allow new lines for all formats. Do you have an opinion for this?

Thanks for all your work.

**Comment by fuhrmanator** [comment]
If you're showing text in HTML, I don't think a newline will break, so you'd have to convert it to `<br/>`. Printing it to console would not require any changes.

After picking my kids up from school, I just realized that Travis-CI Unix failed because of true Unix line endings (WSL-Ubuntu was misleading because I was using the same GIFT test files as Windows - being lazy). So I'm trying to fix those now. It shouldn't change how the previewer works if you're coding in Windows.

**Comment by fuhrmanator** [comment]
> so we should probably allow new lines for all formats. Do you have an opinion for this?

I had the same thought. Finally, I think newlines don't really matter in HTML except for how the HTML is formatted in an editor that shows the HTML (Moodle allows you to toggle that). I believe that's the only reason they appear in GIFT. So, yes, I just leave them as-is.

**Comment by didoudiaz** [comment]
My two cents: in HTML most of the time, newlines are treated as white-spaces except in some contexts like `<pre>...</pre>`, `<code>...</code>`, or in some `<span>...</span>`,... Since, GIFT allows the user to write `[html]` I use newlines (I mean not the litteral `\n` but really entering a newline in the editor) to make the questions more readable and inside blocks `<pre>...</pre>` (or other as explained above). Here is an example of GIFT question about programming (code). 
```
[html]What is returned by mystery(2) ?
<span class\="" style\="font-family\: monospace;color\: blue;white-space\: pre-wrap;">
int mystery(int x) \{
&nbsp;   switch (x) \{
&nbsp;      case 1\: x \= 4;
&nbsp;      case 2\: x \= 5;
&nbsp;      case 3\: x \= 6;
&nbsp;   \}
&nbsp;   return x;
\}
</span>
```
Remarks:

- I use `<span>` to obtain a monospace font because our moodle is configured in such a way it does not display well `<pre>...</pre>`, `<code>...</code>` blocks.
- the `&nbsp;` are because moodle seems to ignore spaces at the begining of a line (at least on our moodle).
- It is a pain, in such a "code" context, to have to protect `{ } = ...` with `\`
- the possibility to write newlines make however the source GIFT question more readable than using a single line with `\n`.

The result is correctly displayed on both our moodle and inside VSCode with the GIFT plugin.

**Comment by fuhrmanator** [comment]
That's a great example @didoudiaz. I added it to the `newline.gift` test.

**Comment by didoudiaz** [comment]
From what you said, I supppose the result of the parser should be similar if I write a single line with several `\n` as in:
```
[html]What is returned by mystery(2) ?\n<span class\="" style\="font-family\: monospace;color\: blue;white-space\: pre-wrap;">\nint mystery(int x) \{\n&nbsp;   switch (x) \{\n&nbsp;      case 1\: x \= 4;\n&nbsp;      case 2\: x \= 5;\n&nbsp;      case 3\: x \= 6;\n&nbsp;   \}\n&nbsp;   return x;\n\}\n</span>
```

**Comment by fuhrmanator** [comment]
Indeed, we should add that test, too ;-)

**Comment by fuhrmanator** [comment]
The HTML formatting was in pretty bad shape wrt newlines and white space on the ends. Looking at this found some bad tests, which I fixed and pushed on the branch.

**Comment by ethan-ou** [comment]
@didoudiaz not related to this topic but might be generally helpful: I'm adding a quick action to the extension that will help you escape things much faster. Just select the block of code you'd like to escape and the extension will handle it for you.

You'll be able to access it through the lightbulb icon (quick actions) or if you select the code, right click > refactor > escape special characters. I'd upload a preview but my internet's going nuts today.

Anyway back to the topic.

**Comment by didoudiaz** [comment]
Great, believe me this will be more than appreciated by a lot of people ! 
How can I update the plugin inside VSCode to test this beta functionnality ?

BTW, do you plan to add other functionalities ? Here are some ideas:

1. replace all real newlines by `\n`. The result is a single (possibly long) line.
2. replace all `\n`by real newlines (the reverse action)
3. In a selected block, replace spaces by `&nbsp;`
4. In a selected block, replace each `&nbsp;` by a space (reverse action)
5. In a selected block, replace a leading space by `&nbsp;` (only if the line starts with a space)
6. indent/desindent a block (move 4 spaces forwards/backwards the block) I don't know if this should be in your plugin or provided by VSCode.

Thank you very much for you help

**Comment by didoudiaz** [comment]
@ethan-ou Thank you very much. I have some suggestions (see in my original issue in your project).

**Comment by fuhrmanator** [comment]
@ethan-ou can we close this issue? 

By the way, I played around with code formatting in GIFT and came up with something that works (maybe @didoudiaz  already found something like this):

```gift
::ArrowFunction equivalent classic function::[markdown]Quel est l'équivalent de\n```\nfunction bonjour(qui\: string) \{\n    return console.log("Bonjour " + qui);\n\}\n```\n?
{
	=`const bonjour \= (qui\: string) \=> console.log("Bonjour " + qui)`
	~`const bonjour \= (qui\: string) \=> return console.log("Bonjour " + qui)`
	~`const bonjour \= qui\: string \=> console.log("Bonjour " + qui)`
	~`(qui\: string) \=> console.log("Bonjour " + qui)`
}
```

It displays in Moodle like (randomized order of answers):

![image](https://user-images.githubusercontent.com/7606540/119208673-4795c300-ba71-11eb-89ec-3d1ea23a459f.png)

The previewers (VSCode and GIFT Question Editor) don't do well with the spaces, but Moodle seems to show it well. So, I'm thinking we can close this issue.

**Comment by ethan-ou** [comment]
@fuhrmanator feel free to close. Also backticks and markdown doesn't look too bad of an idea. I'll definitely pick that up.
